When restoring with lock file, NU1603 warning shouldn't be raised.

We're trying to adopt Locked package files in SDK 2.1.500, and were trying it out which revealed an issue. When we restore using `--locked-mode` (as if by CI), a lot of packages cannot be found, shown in messages like `warning NU1603: System.Globalization 4.0.11 depends on Microsoft.NETCore.Platforms (>= 1.0.1) but Microsoft.NETCore.Platforms 1.0.1 was not found. An approximate best match of Microsoft.NETCore.Platforms 2.1.0 was resolved.`

In this case, I would expect that the same warnings during a regular restore would be present during a `--locked-mode` restore -- in this case, a regular restore has no issues while a locked restore has "hundreds" of these.

#### We're also seeing errors

In our larger solution, we're seeing errors as well, such as the below. This may be related to the fact that we're using GitVersion, so maybe the version on a project changes after the lock file is created.. ?.. That's another issue though.

```
C:\Our-SDK\Evidence\ProjectA\ProjectA.csproj : error NU1004: The packages lock file is inconsistent with the project dependencies so restore can't be run in locked mode. Please disable RestoreLockedMode MSBuild property or pass explicit --force-evaluate flag to run restore to update the lock file. [C:\Our-SDK\Our-SDK.sln]
```

## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe):
Dotnet.exe used for restore, new project, etc.

NuGet version (x.x.x.xxx): `N/A`

dotnet.exe --version (if appropriate): `2.1.500`

OS version (i.e. win10 v1607 (14393.321)): `Win10 10.0.17134`

Worked before? If so, with which NuGet version: `N/A -- new feature in 2.1.500`

## Detailed repro steps so we can see the same problem

1. Create new console project, `dotnet new console`
2. Add a few packages, `dotnet add package Microsoft.EntityFrameworkCore`
3. Restore and create locked file, `dotnet restore --use-lock-file`
4. Restore using locked mode, `dotnet restore --locked-mode`

Now you should see warnings like:

```
warning NU1603: NETStandard.Library 2.0.3 depends on Microsoft.NETCore.Platforms (>= 1.1.0) but Microsoft.NETCore.Platforms 1.1.0 was not found. An approximate best match of Microsoft.NETCore.Platforms 2.1.0 was resolved.

warning NU1603: System.Globalization 4.0.11 depends on Microsoft.NETCore.Targets (>= 1.0.1) but Microsoft.NETCore.Targets 1.0.1 was not found. An approximate best match of Microsoft.NETCore.Targets 2.1.0 was resolved.
```

## Other suggested things

### Verbose Logs

Please include verbose logs (NuGet.exe <COMMAND> -verbosity detailed | dotnet.exe <COMMAND> --verbosity diag | etc...)

`dotnet restore --locked-mode --verbosity diag` => [dotnet.log](https://github.com/NuGet/Home/files/2599940/dotnet.log)

### Sample Project

Very helpful if you can zip a project and paste into this issue!

[bb.zip](https://github.com/NuGet/Home/files/2599943/bb.zip)

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnexpectedDependencyMessages.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnexpectedDependencyMessages.cs
@@ -37,12 +37,17 @@ namespace NuGet.Commands
             ignoreIds.UnionWith(projectMissingLowerBounds.Select(e => e.LibraryId));
             await logger.LogMessagesAsync(DiagnosticUtility.MergeOnTargetGraph(projectMissingLowerBounds));
 
-            // 2. Detect dependency and source issues across the entire graph 
-            //    where the minimum version was not matched exactly.
-            //    Ignore packages already logged by #1
-            var missingMinimums = GetMissingLowerBounds(graphList, ignoreIds);
-            ignoreIds.UnionWith(missingMinimums.Select(e => e.LibraryId));
-            await logger.LogMessagesAsync(DiagnosticUtility.MergeOnTargetGraph(missingMinimums));
+            // Ignore generating NU1603/NU1602 across entire graph if lock file is enabled. Because
+            // lock file enforce a fixed resolved version for all the different requests for the same package ID.
+            if (!PackagesLockFileUtilities.IsNuGetLockFileSupported(project))
+            {
+                // 2. Detect dependency and source issues across the entire graph 
+                //    where the minimum version was not matched exactly.
+                //    Ignore packages already logged by #1
+                var missingMinimums = GetMissingLowerBounds(graphList, ignoreIds);
+                ignoreIds.UnionWith(missingMinimums.Select(e => e.LibraryId));
+                await logger.LogMessagesAsync(DiagnosticUtility.MergeOnTargetGraph(missingMinimums));
+            }
 
             // 3. Detect top level dependencies that have a version different from the specified version.
             //    Ignore packages already logged in #1 and #2 since those errors are more specific.

--- a/test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Commands.Test/UnexpectedDependencyMessagesTests.cs
@@ -779,6 +779,79 @@ namespace NuGet.Commands.Test
             UnexpectedDependencyMessages.DependencyRangeHasMissingExactMatch(dependency).Should().BeFalse("Project type should return false, regardless of the range.");
         }
 
+        [Fact]
+        public async Task ProjectWithoutLockFile_GeneratesNU1603()
+        {
+            var testLogger = new TestLogger();
+            var range = VersionRange.Parse("2.0.0");
+            var tfi = GetTFI(NuGetFramework.Parse("net46"), new LibraryRange("x", range, LibraryDependencyTarget.Package));
+            var project = new PackageSpec(tfi)
+            {
+                Name = "proj"
+            };
+            var flattened = new HashSet<GraphItem<RemoteResolveResult>>
+            {
+                new GraphItem<RemoteResolveResult>(new LibraryIdentity("X", NuGetVersion.Parse("2.0.0"), LibraryType.Package))
+            };
+            var targetGraph = new Mock<IRestoreTargetGraph>();
+            targetGraph.SetupGet(e => e.Flattened).Returns(flattened);
+            targetGraph.SetupGet(e => e.TargetGraphName).Returns("net46/win10");
+            targetGraph.SetupGet(e => e.Framework).Returns(NuGetFramework.Parse("net46"));
+            var parent = new LibraryIdentity("z", NuGetVersion.Parse("9.0.0"), LibraryType.Package);
+            var child = new LibraryIdentity("x", NuGetVersion.Parse("2.0.0"), LibraryType.Package);
+            var dependency = new ResolvedDependencyKey(parent, VersionRange.Parse("1.0.0"), child);
+            var dependencySet = new HashSet<ResolvedDependencyKey>() { dependency };
+            targetGraph.SetupGet(e => e.ResolvedDependencies).Returns(dependencySet);
+            var targetGraphs = new[] { targetGraph.Object };
+            var ignore = new HashSet<string>();
+
+            await UnexpectedDependencyMessages.LogAsync(targetGraphs, project, testLogger);
+
+            testLogger.LogMessages.Select(e => e.Code).Should().NotContain(NuGetLogCode.NU1604);
+            testLogger.LogMessages.Select(e => e.Code).Should().NotContain(NuGetLogCode.NU1601);
+            testLogger.LogMessages.Select(e => e.Code).Should().Contain(NuGetLogCode.NU1603);
+        }
+
+        [Fact]
+        public async Task ProjectWithLockFile_NU1603_NotGenerated()
+        {
+            var testLogger = new TestLogger();
+            var range = VersionRange.Parse("2.0.0");
+            var tfi = GetTFI(NuGetFramework.Parse("net46"), new LibraryRange("x", range, LibraryDependencyTarget.Package));
+            var project = new PackageSpec(tfi)
+            {
+                Name = "proj",
+                RestoreMetadata = new ProjectRestoreMetadata()
+                {
+                    RestoreLockProperties = new RestoreLockProperties(
+                        restorePackagesWithLockFile: "true",
+                        nuGetLockFilePath: null,
+                        restoreLockedMode: true)
+                }
+            };
+            var flattened = new HashSet<GraphItem<RemoteResolveResult>>
+            {
+                new GraphItem<RemoteResolveResult>(new LibraryIdentity("X", NuGetVersion.Parse("2.0.0"), LibraryType.Package))
+            };
+            var targetGraph = new Mock<IRestoreTargetGraph>();
+            targetGraph.SetupGet(e => e.Flattened).Returns(flattened);
+            targetGraph.SetupGet(e => e.TargetGraphName).Returns("net46/win10");
+            targetGraph.SetupGet(e => e.Framework).Returns(NuGetFramework.Parse("net46"));
+            var parent = new LibraryIdentity("z", NuGetVersion.Parse("9.0.0"), LibraryType.Package);
+            var child = new LibraryIdentity("x", NuGetVersion.Parse("2.0.0"), LibraryType.Package);
+            var dependency = new ResolvedDependencyKey(parent, VersionRange.Parse("1.0.0"), child);
+            var dependencySet = new HashSet<ResolvedDependencyKey>() { dependency };
+            targetGraph.SetupGet(e => e.ResolvedDependencies).Returns(dependencySet);
+            var targetGraphs = new[] { targetGraph.Object };
+            var ignore = new HashSet<string>();
+
+            await UnexpectedDependencyMessages.LogAsync(targetGraphs, project, testLogger);
+
+            testLogger.LogMessages.Select(e => e.Code).Should().NotContain(NuGetLogCode.NU1604);
+            testLogger.LogMessages.Select(e => e.Code).Should().NotContain(NuGetLogCode.NU1601);
+            testLogger.LogMessages.Select(e => e.Code).Should().NotContain(NuGetLogCode.NU1603);
+        }
+
         [Theory]
         [InlineData("(1.0.0, )")]
         [InlineData("(, 1.0.0]")]

